src/bitcoind.cpp

```diff
@@ -28,6 +29,79 @@
 const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
 UrlDecodeFn* const URL_DECODE = urlDecode;
 
+#if HAVE_DECL_FORK
+
+/** Custom implementation of daemon(). This implements the same order of operations as glibc.
+ * Opens a pipe to the child process to be able to wait for an event to occur.
+ *
+ * @returns 0 if succesful, and in child process.
+ *          >0 if succesful, and in parent process.
+ *          -1 in case of error (in parent process).
+ *
+ *          In case of success, endpoint will be one end of a pipe from the child to parent process,
+ *          which can be used with TokenWrite (in the child) or TokenRead (in the parent).
+ */
+int fork_daemon(bool nochdir, bool noclose, TokenPipeEnd &endpoint)
+{
+    // communication pipe with child process
+    std::optional<TokenPipe> umbilical = TokenPipe::create();
+    if (!umbilical) {
+        return -1; // pipe or pipe2 failed.
+    }
+
+    int pid = fork();
+    if (pid < 0) {
+        return -1; // fork failed.
+    }
+    if (pid != 0) {
+        // Parent process gets read end, closes write end.
+        endpoint = umbilical->ReadEnd();
+        umbilical->WriteEnd();
+
+        int status = endpoint.TokenRead();
+        if (status != 0) { // Something went wrong while setting up child process.
+            endpoint.Close();
+            return -1;
+        }
+
+        return pid;
+    }
+    // Child process gets write end, closes read end.
+    endpoint = umbilical->WriteEnd();
+    umbilical->ReadEnd();
```

Comment:
Calling `umbilical->ReadEnd()` to create the `TokenPipeEnd` then immediately destroy it, which then closes the underlying fd seems a bit clunky. A comment might be worthwhile at least.

Reply:
I thought about adding an explicit `CloseReadEnd` `CloseWriteEnd` function but as it'd essentially be the same I didn't do so. Will add a comment.

Reply:
What if I call the functions `TakeReadEnd` `TakeWriteEnd` would that make it clearer that a move is happening?

Reply:
Could also write `umbilical->ReadEnd().Close();` (relying on the destructor's second call to Close then being a no-op). I don't really have a preference.

Reply:
Ah yea! I think I'll do both.
```
umbilical->TakeReadEnd().Close();
```
Fully self-documenting then.